chia/wallet/util/puzzle_compression.py

```diff
@@ -0,0 +1,203 @@
+"""
+
+HASH_TO_DRIVER: Dict[uint16, Dict[bytes32, Any]] = {
+    uint16(0): {
+        standard_puzzle.MOD.get_tree_hash(): StandardPuzzle,
+        CC_MOD.get_tree_hash(): CATPuzzle,
+        OFFER_MOD.get_tree_hash(): OfferPuzzle,
+    }
+    # TODO: Add version tests when there are more versions :)
+}
+
+LATEST_VERSION: uint16 = uint16(max(HASH_TO_DRIVER.keys()))
+
+
+class CompressionVersionError(Exception):
+    pass
+
+
+class PuzzleCompressor:
+    """
+    This class represents an instance of the global HASH_TO_DRIVER map above.
+
+    You can either specify a version number which will use the identifiers that are defined up to that version,
+    or you can specify an entirely custom dictionary of drivers that is not a valid version.
+    """
+
+    version_number: uint16
+    driver_dict: Dict[bytes32, Any]
+
+    def __init__(self, version: uint16 = LATEST_VERSION, driver_dict: Dict[bytes32, Any] = None):
+        if driver_dict is None:
+            assert version in HASH_TO_DRIVER.keys()
+            self.version_number = uint16(version)
+            final_dict: Dict[bytes32, Any] = {}
+            for key in range(0, version + 1):
+                final_dict = final_dict | HASH_TO_DRIVER[uint16(key)]
+            self.driver_dict = final_dict
+        else:
+            self.version_number = uint16(65535)  # Just set it to max so it always passes any version checks
+            self.driver_dict = driver_dict
+
+    def match_puzzle(self, puzzle: Program) -> Tuple[bool, Union[PuzzleRepresentation, Program]]:
+        for identifier, driver in self.driver_dict.items():
+            matched, args = driver.match(puzzle, self)
+            if matched:
+                return True, PuzzleRepresentation(identifier, args)
+        return False, puzzle
+
+    def serialize(self, puzzle: Program) -> bytes:
+        _, rep = self.match_puzzle(puzzle)
+        return bytes(self.version_number) + bytes(rep)
+
+    def deserialize(self, object_bytes: bytes) -> Program:
+        if int.from_bytes(object_bytes[0:2], "big") > self.version_number:
+            raise CompressionVersionError()
+        else:
+            deversioned_bytes = object_bytes[2:]
+            try:
+                program = PuzzleRepresentation.from_bytes(deversioned_bytes).construct(self.driver_dict)
+            except Exception:
+                program = Program.from_bytes(deversioned_bytes)
+            return program
+
+    @classmethod
+    def get_identifier_version(cls, identifier: bytes32) -> "PuzzleCompressor":
+        for id, driver_dict in HASH_TO_DRIVER.items():
+            if identifier in driver_dict:
+                return cls(id)
+        raise ValueError("The given identifier does not have a puzzle driver mapping")
+
+    @classmethod
+    def lowest_compatible_version(
+        cls, matched_puzzles: List[Union[PuzzleRepresentation, Program]]
+    ) -> "PuzzleCompressor":
```

Comment:
This will be an incorrect hint when someone subclasses `PuzzleCompressor`.  And it would be nice to not inherit, but we do.  This applies multiple places.  Maybe the typevar shouldn't be private, I dunno.

Add above:

```python
_T_PuzzleCompressor = TypeVar("_T_PuzzleCompressor", bound="PuzzleCompressor")
```

```suggestion
    @classmethod
    def lowest_compatible_version(
        cls: Type[_T_PuzzleCompressor], matched_puzzles: List[Union[PuzzleRepresentation, Program]]
    ) -> _T_PuzzleCompressor:
```